Refactor/make Swipeable component

enejb opened the pull request:
This adds a swipeable component and integrates it into the home section.  

#### Changes proposed in this Pull Request

This PR refactors https://github.com/Automattic/wp-calypso/pull/58533
by extracting a new swipeable component. That enables horizontal swiping. 

#### Testing instructions

Visit /home/example.com 
Notice that you are able to swipe the 2 dot components are now as expected.
You are also able to use the arrows and the dot components (slideshows) behave as before.

Visit /devdocs/design/swipeable
and you are able to swipe though the different pages as expected.

Visit /devdocs/design/dot-pager
and you are able to swipe through the different pages as expected.

Try this across different browsers. 
- iOS Simulator ( Launch xcode )  - you are not able to login to calypso.localhost or any other site other the wordpress.com - but you can test it using the devdocs url.
- Firefox, Chrome, Edge, Safari 

Try it also using the dev tools and resize the window to different widths. 

See

https://user-images.githubusercontent.com/115071/145516869-990e36c5-236a-47a4-a19a-5b96d0a0270b.mp4

enejb committed:
Add touch controls.

enejb committed:
Rebase

enejb committed:
Use react resize aware

enejb committed:
Revert "Use react resize aware"

This reverts commit 8a2999c57befcf62b550f5e8559f0d7669534c3f.

enejb committed:
Use resize observer element that we can use to dynamically calculate the width

enejb committed:
Remove the display: table since it breaks some examples.

enejb committed:
Fix for leaving the draggable area.

enejb committed:
Add: Swipeabel component

enejb committed:
Convert indentation to tabs

enejb committed:
Lint fixes to readme.

tyxla commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,212 @@
+
+import './style.scss';
+
+export const Swipable = ( {
```
Seems like we have a typo:
```suggestion
export const Swipeable = ( {
```

tyxla commented on client/components/swipeable/README.md:
```diff
@@ -0,0 +1,33 @@
+# Support Info
```
```suggestion
# Swipeable
```

tyxla commented on client/components/swipeable/README.md:
````diff
@@ -0,0 +1,33 @@
+## Example Usage
+
+```js
+import Swipable from 'calypso/components/swipable';
````
Could we fix all the typos where we have `swipable` instead of `swipeable`? There are a few more in the PR.

tyxla commented on client/components/swipeable/README.md:
```diff
@@ -0,0 +1,33 @@
+				setCurrentPage( index );
+			} }
+			currentPage
+			pageClassName=""
```
Should we provide some value here? If we leave it empty, there's a high chance someone will use it like that in their code.

ciampo replied:
Agreed, both for `currentPage` and for `pageClassName`

tyxla commented on client/components/swipeable/README.md:
```diff
@@ -0,0 +1,33 @@
+
+## Props
+
+- `onPageSelect` - (function) Function that .
```
Needs to be finished.

tyxla commented on client/components/swipeable/README.md:
```diff
@@ -0,0 +1,33 @@
+## Props
+
+- `onPageSelect` - (function) Function that .
+- `pageClassName` - _optional_ (string) A URL leading to an overview of a feature. If `false`, no link will be displayed.
```
Description here seems incorrect.

tyxla commented on client/components/swipeable/README.md:
```diff
@@ -0,0 +1,33 @@
+- `onPageSelect` - (function) Function that .
+- `pageClassName` - _optional_ (string) A URL leading to an overview of a feature. If `false`, no link will be displayed.
+- `currentPage` - (int) Index of a selected page.
+- `hasDynamicHeight` - (bool) should height by dynamically changed.
```
There seems to be a grammatic issue in this one:

```suggestion
- `hasDynamicHeight` - (bool) Whether height should be changed dynamically
```

Also, what exactly does "Dynamic" mean? Does it need some sort of further clarification?

tyxla reviewed (commented):
Thanks for extracting it!

I'm leaving this to @Automattic/team-calypso-components for an expert review, but I've left some small suggestions in the meantime.

ciampo commented on client/components/swipeable/README.md:
```diff
@@ -0,0 +1,33 @@
+
+function Pager() {
+	const [ currentPage, setCurrentPage ] = useState( 0 );
+	return ( 
```
Nit: trailing whitespace

```suggestion
	return (
```

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,212 @@
+	children,
+	currentPage = 0,
+	onPageSelect,
+	pageClassName,
```
Should we get any other props via the spread syntax (i.e `...props`) and then pass them onto the outer `div` rendered below?

enejb replied:
I think we can always add this later but I don't think it's needed at this point.

ciampo replied:
It can be added later — I just think it's good practice for a component to forwards its props, and I thought it would be a minor change that could be added to this PR.

enejb replied:
Sorry I didn't realize that it was a good practice. Added.

ciampo commented on client/components/dot-pager/index.jsx:
```diff
@@ -83,76 +84,36 @@ export const DotPager = ( {
-				} }
+				setCurrentPage={ handleSelectPage }
 			/>
-			<div className="dot-pager__pages" ref={ pagesRef } style={ pagesStyle }>
```
If any consumer of this component was relying on the `dot-pager__pages`, then removing it may be a breaking change.

Maybe it could be passed as a prop to `Swipeable`, which would then apply it to its outer `div` ?

enejb replied:
Good idea. I wasn't able to find an instance of the `dot-pager__pages` but I guess the components could be used somewhere else.

ciampo commented on client/components/dot-pager/index.jsx:
```diff
@@ -83,76 +84,36 @@ export const DotPager = ( {
 
+	const handleSelectPage = ( index ) => {
+		setCurrentPage( index );
+		onPageSelected && onPageSelected( index );
```
Just a syntax preference, feel free to ignore

```suggestion
		onPageSelected.?( index );
```

enejb replied:
THis new syntax didn't work for me.

ciampo replied:
Interesting, would be curious to know what didn't work

tyxla replied:
Me too, actually! In this instance, a function call with optional chaining should work the same way as with the boolean evaluation.

enejb replied:
This is what I see in the editor. 
<img width="603" alt="Screen Shot 2021-12-02 at 3 28 28 PM" src="https://user-images.githubusercontent.com/115071/144519417-7cb0a492-ff56-4946-a209-03cf5330bf92.png">

ciampo replied:
🤦  My bad, sorry. The correct syntax would be

```suggestion
		onPageSelected?.( index );
```

(swapping . and ?)

tyxla replied:
@enejb here's some more information about the optional chaining syntax: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

ciampo commented on client/components/swipeable/style.scss:
```diff
@@ -0,0 +1,32 @@
+@import '@wordpress/base-styles/mixins';
+.swipeable__container {
+	width: 100%;
+	height: 100%;
+	margin-left: auto;
+	margin-right: auto;
```
It is usually a good practice for a component to stop at its `border-box`, and let the component's consumer set the occupied space and especially the margin behaviour.

I would suggest moving the `margin-*` rules (and potentially the `height` too) to where this component is consumed (i.e. the Dot Pager component)

enejb replied:
Good idea. I removed the margin-* properties.

ciampo commented on client/components/swipeable/style.scss:
```diff
@@ -0,0 +1,32 @@
+	width: 100%;
+	flex-shrink: 0;
+	margin-left: 0;
+}
```
Nit: adding a trailing empty line
```suggestion
}

```

ciampo commented on client/devdocs/style.scss:
```diff
@@ -155,7 +155,6 @@ $devdocs-max-width: 720px;
 		// A wrapper inside a wrapper to help reliably center
 		// component and block examples.
 		.docs-example__wrapper-content-centering {
-			display: table;
```
Why is this change necessary? It doesn't seem related to Dot Pager or Swipeable

enejb replied:
The reason why it is necessary is that the dot pages component of the dev docs doesn't work as expected when `display: table` is set here.

ciampo replied:
Got it. We should make sure that this change doesn't introduce a regression for other components that were potentially relying on this CSS rule for their docs.

enejb replied:
I tested 
[/devdocs/design](http://calypso.localhost:3000/devdocs/design) and 
[/devdocs/blocks](http://calypso.localhost:3000/devdocs/blocks) and I wasn't able to notice anything broken.

ciampo commented on client/components/swipeable/style.scss:
```diff
@@ -0,0 +1,32 @@
+	display: flex;
+
+	transition: all 0.5s ease-in-out;
+	transition-property: transform, height;
```
Usually transitioning the `height` property is not recommended, as it can cause poor rendering performance due to the fact that every change in `height` can cause layout / reflow

enejb replied:
This is mostly there to be backwards compatible with how dot pages currently work.

ciampo replied:
Got it. Have you considered removing the transition entirely for the `height` property?

enejb replied:
I did but then we don't see the height change effect that is currently present on /home/example.com 
Instead, the height doesn't change at all and there is a big whitespace at the bottom of the page.

ciampo replied:
I see. Let's leave this for now, we can always change it later if needed

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,212 @@
+	const resizeObserverRef = useRef();
+	const numPages = Children.count( children );
+
+	const getWidth = () => {
+		return resizeObserverRef.current?.getBoundingClientRect().width;
+	};
```
Calling `getBoundingClientRect` can be quite expensive, we should probably look changing this from being a function that is called many times per render, into a memoized value which is recalculated only when necessary

enejb replied:
I added the calculation and stored the value in the state

ciampo commented on client/components/swipeable/style.scss:
```diff
@@ -0,0 +1,32 @@
+	overflow: hidden;
+	list-style: none;
+	padding: 0;
+	z-index: 1;
```
Why is the `z-index` necessary?

enejb replied:
I thought it would help but I don't think we need it.

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,212 @@
+			return; // End early if we are not dragging any more.
+		}
+
+		const THRESHOLD_OFFSET = 100; // Number of pixels to travel before we trigger the slider to move to the desired slide.
```
Should this be defined at the very top of the file (after the imports), as a constant?

enejb replied:
I like when constants are defined where they are used when they are only used once. 
But if you want to see this change I am open to it.  :)

ciampo replied:
My main point about this is that this constant would be re-defined every time the `handleDragEnd` function runs.

Anyway it's definitely not a big deal, feel free to keep it as you prefer.

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,212 @@
+			transitionDuration: `300ms`,
+		} );
+		onPageSelect( newIndex );
+		setDragStartData( {} );
```
Why using an empty object instead of `null` or `undefined` ? (Same question for this state's initial value)

enejb replied:
Good Idea. I was able to remove the isDragging state as well.

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,212 @@
+			} );
+		}
+
+		const swipeableArea = pagesRef.current?.getBoundingClientRect();
```
Similarly to what state above, calling `getBoundingClientRect` on every drag event can hurt rendering performance. Ideally the size of the swipeable area should be a memoized value that is recalculated only when necessary

enejb replied:
Good call!

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,212 @@
+		setIsDragging( false );
+	};
+
+	const handleDrag = ( event ) => {
```
This may not be as important as other perf-related comments, but we should look into memoizing (via the `useCallback` hook) these functions, which are otherwise added/removed as event handlers for every render:

- `getDragPosition`
- `handleDragStart`
- `handleDrag`
- `handleDragEnd`

enejb replied:
From my limited performance testing, all the animations seem to run well. (58 fps) I am not sure we need to add the useCallback hooks.

ciampo replied:
This could definitely be a bit of premature optimisation, but since you're mentioning FPS, what device did you test it on?

I'm asking because for this kind of testing, we should make sure that we can hit (ideally) 60fps on a low-end device — testing on the machines that we use to code daily is not a good reference for the average use (cc @sgomes )

sgomes replied:
It's always a good idea to measure performance and make sure that you're making an informed decision, and as @ciampo points out, doing so on lower-end devices that we want to target as our baseline (or, alternatively, using something like the CPU slowdown feature in Chrome DevTools). This avoids needlessly complicating the code, but making sure that our bases are well covered.

That said, adding and removing event handlers on every render is probably not a great practice, even if React has its own event handling layer that sits atop the DOM. I'm a bit rusty on my React hooks, but if this is achievable with minimal code and memory overhead, it's probably a good idea to do a little less work on every render.

Do let me know if you'd like me to dig deeper and do some runtime performance testing on this!

ciampo reviewed (commented):
Happy to give this PR a look, although I haven't been working in Calypso in a while and I may lack context (about the dot pager and its usage, about general conventions in the Calypso client's components...) — so it would be definitely better if someone familiar with the related project also gave it a look.

Anyway, apart from the inline comments, I also have a few more general thoughts:

- is there a way to see the `Swipeable` component in isolation? It can help a lot when reviewing component-related changes
- should we add any unit tests?
- Is there any CHANGELOG to update?

enejb committed:
fix typo

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

enejb committed:
fix typo

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

enejb committed:
remove extra space

Co-authored-by: Marco Ciampini <marco.ciampini@automattic.com>

enejb committed:
update to prefered syntax

Co-authored-by: Marco Ciampini <marco.ciampini@automattic.com>

enejb committed:
Add new line at the end of the file.

Co-authored-by: Marco Ciampini <marco.ciampini@automattic.com>

enejb committed:
Update title or readme

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

enejb committed:
revert linter was complaining

enejb committed:
Fix all instances of swipeable typo

enejb committed:
Update Readme with better examples and explantation

enejb committed:
More updates to the readme.

enejb committed:
Add doc example

To test visit /devdocs/design/swipeable

enejb committed:
Remove the not needed styles from swipeable

enejb committed:
Store pageWidth in the state instead of calculating it everytime.

enejb committed:
Remove isDragging state, just use dragStartData instead

enejb committed:
Add hasDynamicHeight to examples.

enejb committed:
memorize swipeableArea in state

enejb committed:
Add velocity treshold for swiping

enejb committed:
Add support for isRTL

enejb committed:
Fix NaN error

enejb committed:
Don't asign duplicate events

enejb committed:
Added containerClassName for backwards compatibility

enejb commented:
@tyxla and @ciampo - Thank you so much for all the feedback. I think I addressed it all.

@ciampo - The `dot-pager` component is only currently used in the /home page

One change that I want to make is removing `hasDynamicHeight={ true }` I think it should always be set to true. 
I tested the changes on iOS safari simulator, Firefox and Chrome and it worked as expected for me. 

I tested it both on an RTL language and regular. 

Let me know if you want to see any other changes here.

tyxla commented on client/components/dot-pager/docs/example.jsx:
```diff
@@ -2,7 +2,7 @@ import DotPager from '../';
 export default function DotPagerExample() {
 	return (
-		<DotPager>
+		<DotPager hasDynamicHeight={ true }>
```
`={ true }` is redundant.

```suggestion
		<DotPager hasDynamicHeight>
```

tyxla commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,249 @@
+			return;
+		}
+
+		// Did the user swipe out of the swipable area?
```
Typo here:

```suggestion
		// Did the user swipe out of the swipeable area?
```

tyxla commented on client/components/swipeable/README.md:
```diff
@@ -0,0 +1,34 @@
+- `onPageSelect` - (function) Function that runs when the page is selected on using a swipe. Useful for updating the current page.
+- `currentPage` - (int) Index of the currently selected/shown page.
+- `hasDynamicHeight` - (bool) Whether height should be changed dynamically.
+- `pageClassName` - _optional_ (string) Optional class attrbitue of the page component.
```
Typo here:

```suggestion
- `pageClassName` - _optional_ (string) Optional class attribute of the page component.
```

tyxla commented on client/components/swipeable/README.md:
```diff
@@ -0,0 +1,34 @@
+- `currentPage` - (int) Index of the currently selected/shown page.
+- `hasDynamicHeight` - (bool) Whether height should be changed dynamically.
+- `pageClassName` - _optional_ (string) Optional class attrbitue of the page component.
+- `containerClassName` - _optional_ (string) Optional class attrbitue of the page container component.
```
```suggestion
- `containerClassName` - _optional_ (string) Optional class attribute of the page container component.
```

tyxla commented on client/components/swipeable/docs/example.jsx:
```diff
@@ -0,0 +1,28 @@
+import { PureComponent } from 'react';
+import Swipeable from 'calypso/components/swipeable';
+
+export default class extends PureComponent {
```
I think we can make this a functional component? We're usually refactoring class components when they don't need to be class components. And people tend to copy a lot from the examples 😉

enejb replied:
Good idea.

tyxla commented on client/components/swipeable/docs/example.jsx:
```diff
@@ -0,0 +1,28 @@
+	constructor( props ) {
+		super( props );
+		this.state = { currentPage: 0 };
+	}
```
If we decided to use a class component, `state` could be declared as a class instance property, and we wouldn't need to have a constructor:

```suggestion
	state = {
		currentPage: 0,
	};
```

tyxla commented on client/components/swipeable/docs/example.jsx:
```diff
@@ -0,0 +1,28 @@
+				} }
+				currentPage={ this.state.currentPage }
+				pageClassName="example-page-component-class"
+				hasDynamicHeight={ true }
```
```suggestion
				hasDynamicHeight
```

tyxla commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,249 @@
+	};
+
+	return (
+		<div>
```
Do we need that `div`? Could it be a `<>...</>`?

enejb replied:
Good call.

tyxla commented on client/devdocs/design/index.jsx:
```diff
@@ -190,6 +191,7 @@ export default class DesignAssets extends Component {
 					<DateRange readmeFilePath="date-range" />
 					<DiffViewerExample readmeFilePath="diff-viewer" />
 					<DotPager readmeFilePath="dot-pager" />
+					<Swipeable readmeFilePath="swipeabel" />
```
The readme won't be found because we have a typo:
```suggestion
					<Swipeable readmeFilePath="swipeable" />
```

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,249 @@
+		transform: `translate3d(0px, 0px, 0px)`,
+		transitionDuration: `300ms`,
+	} );
+	const [ dragStartData, setDragStartData ] = useState();
```
With the recent changes, the initial value should also be `null` for consistency

```suggestion
	const [ dragStartData, setDragStartData ] = useState( null );
```

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,249 @@
+		// fire when the window resizes
+		useEffect( () => {
+			if ( ! enabled ) {
+				return;
```
Wouldn't this early return prevent the event listeners from being cleared? We should probably do instead:

```
if ( enabled ) {
  // add ev listeners
}

return {
  if (enabled) {
    // remove event listeners
  }
}

enejb replied:
Can you explain a bit more when this would be the case?

ciampo replied:
Nevermind my initial comment, I misread the code.

The `useEffect` hook returns a [cleanup function](https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect) — I initially thought that, by having an early return, we would miss running the cleanup callback (the one that removes the `resize` and `orientationchange` ev listeners).

But actually, by returning early, we also don't add any of these ev. listeners — and therefore there's nothing to cleanup.

Going to mark this as resolved, sorry!

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,249 @@
+		return isRtl ? offset : -offset;
+	};
+
+	function useUpdateLayout( enabled, currentPageIndex, updateLayout ) {
```
I'm a bit confused by the `useUpdateLayout` function, maybe @tyxla can also help out in case I'm missing something:

- it looks like this function is supposed to be used like a React hook, but it is declared inside the render function of `Swipeable` — should it not be defined outside this component?
- I'm not sure I get why we need to use a `ref` to store/invoke the `updateLayout` callback — wouldn't it be easier to wrap that callback in a `useCallbak` hook where it is defined, instead of assigning it to a ref? It makes this code much harder to follow

tyxla replied:
> it looks like this function is supposed to be used like a React hook, but it is declared inside the render function of Swipeable — should it not be defined outside this component?

I agree, it should be defined outside of the component.

> I'm not sure I get why we need to use a ref to store/invoke the updateLayout callback — wouldn't it be easier to wrap that callback in a useCallbak hook where it is defined, instead of assigning it to a ref? It makes this code much harder to follow

I think what Marco's suggesting makes sense. This likely came to be like that because of an earlier suggestion by @jsnajdr in another PR.

enejb replied:
yes this part of the code is there because of https://github.com/Automattic/wp-calypso/pull/58377

ciampo replied:
> I'm not sure I get why we need to use a ref to store/invoke the updateLayout callback — wouldn't it be easier to wrap that callback in a useCallbak hook where it is defined, instead of assigning it to a ref? It makes this code much harder to follow

Regarding this comment, this is what I mean:

```js
const updateLayoutCallback = useCallback(
  ( hasTransitionDuration = true ) => { /* function body */ },
  [ /* dependencies */ ]
);

useUpdateLayout( updateEnabled, currentPage, updateLayoutCallback );
```

Since now the callback is memoized, you could simplify `useUpdatedLayout` and remove the `ref` — something like

```js
function useUpdateLayout( enabled, currentPageIndex, updateLayout ) {
	// fire when the `currentPageIndex` parameter changes
	useEffect( () => {
		if ( ! enabled ) {
			return;
		}

		updateLayout();
	}, [ enabled, currentPageIndex, updateLayout ] );

	// fire when the window resizes
	useEffect( () => {
		if ( ! enabled ) {
			return;
		}

		const onResize = () => updateLayout( false );
		window.addEventListener( 'resize', onResize );
		window.addEventListener( 'orientationchange', onResize );
		return () => {
			window.removeEventListener( 'resize', onResize );
			window.removeEventListener( 'orientationchange', onResize );
		};
	}, [ enabled, updateLayout ] );
}
```

I haven't tested it on my machine, but I hope it makes sense!

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,249 @@
+		if ( targetHeight && pagesStyle?.height !== targetHeight ) {
+			height = { height: null };
+		}
+		setPageWidth( getWidth() );
+		setSwipeableArea( pagesRef.current?.getBoundingClientRect() );
```
I understand that these variables (that are expensive to compute) and now only recalculated when this callback is run by `useUpdatedLayout` — but does it mean that, if `updateEnabled` is `false`, these values never get computed?

ciampo replied:
@sgomes I wonder if you also have any suggestion for those `getBoundingClientRect` calls — I know they can be quite expensive to recalculate.

I initially suggested ([here](https://github.com/Automattic/wp-calypso/pull/58690#discussion_r760468419) and [here](https://github.com/Automattic/wp-calypso/pull/58690#discussion_r760457768)) to memoize them, but in the meantime they have been moved to this callback.

Do you think this change is reasonable?

sgomes replied:
Happy to help! This needs a little bit of extra background to explain thoroughly.

The browser rendering pipeline is roughly divided into three stages: Layout, Paint, and Composite. Very simply put, the first stage determines the general layout of the page in terms of which elements go where, the second stage renders everything into layers, and the final stage takes those layers and combines them into a single layer to be rendered on the user's screen.

React sits atop of this, having its own "Render" stage, which essentially determines the changes to be applied to the real browser DOM before any of the above begins.

The challenge is that `getBoundingClientRect` triggers a layout pass, which, if it happens at the wrong moment, can throw the whole pipeline out of whack, essentially restarting it and delaying rendering beyond the difficult to hit 13ms (1000ms/60fps) target. Namely, we need to avoid taking these measurements as we're getting ready to render on screen.

It's not currently clear to me at what stage this hook is getting called, nor what's driving it, but the general approach inside of it seems perfectly reasonable to me, and follows what is generally the best practice: first, take all your measurements with the potentially heavy layout-hitting methods you need, and only then start the animation. If this is done well, then even if you miss some frames, no animation will have started yet, so it isn't visible to the user or in any way problematic (assuming it doesn't take a user-noticeable delay to start the animation, but you'd need >50ms for that). That is, you make sure to only run Layout as part of the initial setup before you start animating, and then just let the animation run on its own; the animation in question here is a `transform`, which is a CSS property that only triggers the Composite phase on every frame, which is exactly what we want!

As such, assuming that this is being called once, at the beginning of the process, and that it's not being called at the wrong time in the rendering pipeline, this approach should work well. I'd need to spend a bit of time refreshing my React hook knowledge in order to figure out whether this is the case, though 😕

ciampo replied:
Thank you for the rendering pipeline refreshed, Sergio!

I'll let @enejb go through the other comments in this PR (which may also affect how this callback is run), and we may resume this conversation later

enejb replied:
Thanks for the explenation @sgomes. 
 I tested this with FF and Chrome using 6x CPU slowing and I didn't notice I am still seeing about 58fps any difference between when we did have a useCallback and not. As far as I can tell they don't make any difference. 

I am leaving the change in here. So it might be easier to compare :) if someone else wants to try.

ciampo replied:
Thank you @sgomes and @enejb , we can probably then mark this conversation resolved.

enejb committed:
fix typo

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

enejb committed:
Fix typo

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

enejb committed:
Fix typo

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

enejb committed:
Simplify code

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

enejb committed:
fix typo

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

enejb committed:
Make it more consistant

Co-authored-by: Marco Ciampini <marco.ciampini@automattic.com>

enejb committed:
Return null instead of 0 for width css in getPagesWidth

enejb committed:
Fix syntext to be more concise

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

enejb committed:
Simply the DOM structure

enejb committed:
Add back the cancel events

enejb committed:
Example component convert to functional component

enejb committed:
Add currentPage to setCurrent Page useEffect

enejb committed:
Add other props to the main swipeable div

enejb committed:
Move constants up so that we don't have to define them on render always.

enejb committed:
Move useUpdateLayout outside the component

enejb committed:
Adding useCallback everywhere and move functions out of the component.

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,296 @@
+		if ( targetHeight && pagesStyle?.height !== targetHeight ) {
+			height = { height: null };
+		}
+		setPageWidth( getWidth() );
```
Since the `getWidth()` function is only invoked here, we could also remove that function and replace this line with:

```suggestion
		setPageWidth( resizeObserverRef.current?.getBoundingClientRect().width );
```

ciampo reviewed (commented):
The great majority of the comments on my end have been addressed (only a couple still open), @tyxla would you mind also giving this a final look?

Thank you!

tyxla commented on client/components/swipeable/docs/example.jsx:
```diff
@@ -0,0 +1,24 @@
+			pageClassName="example-page-component-class"
+			hasDynamicHeight
+		>
+			<div>Page 1</div>
```
Optioinal, but I think it would be useful to have a left arrow and right arrow (when available) in the examples. It would make the example more clear as to how it's expected to work.

ciampo replied:
Has this comment been addressed? From the markup I'm not able to see changes that suggest that arrows were added

tyxla commented on client/components/swipeable/docs/example.jsx:
```diff
@@ -0,0 +1,24 @@
+import Swipeable from 'calypso/components/swipeable';
+
+const SwipeableExample = () => {
+	const [ currentPage, setCurrentPage ] = useState( 0 );
```
I've noticed an odd behavior of the example:

* Go to `/devdocs/design`
* Search for "Swipeable"
* Click on the result example.
* You'll end up on `/devdocs/design/swipeable`
* However, you can see all 3 slides on the same page:

![](https://cldup.com/bCQrIAtZP4.png)

If I refresh `/devdocs/design/swipeable`, it all looks and works well.

enejb replied:
I am able to replicate this as well. I think it has to do with the memoize function storing the width of the elements. So when you go to one that has a smaller width than the current one you end up in a situation where the size is not the same anymore. In order to fix we would need a way to force the re-render of the swipeable component. Or have the component listen to some sort of event that would trigger a rerender. Maybe something like https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API ? Or have a prop just for cases like this.

tyxla replied:
The reason is that the container width between the "components list view" and "component single view" is different, so I think this indicates that we're not properly listening to dimensions changes. What if this happens in a real-world scenario? I think we should fix this. Perhaps we could use the [Resize Observer API](https://developer.mozilla.org/en-US/docs/Web/API/Resize_Observer_API)? It is supported in all major browsers that we support AFAIK.

tyxla commented on client/components/swipeable/README.md:
```diff
@@ -0,0 +1,34 @@
+## Props
+
+- `onPageSelect` - (function) Function that runs when the page is selected on using a swipe. Useful for updating the current page.
+- `currentPage` - (int) Index of the currently selected/shown page.
```
Nit: JS has `number`, not `int` 😉

enejb replied:
Should be fixed

tyxla commented on client/components/swipeable/style.scss:
```diff
@@ -0,0 +1,29 @@
+	transition-property: transform, height;
+	@include reduce-motion( 'transition' );
+}
+.swipeable__resize-observer {} /* stylelint-disable-line block-no-empty */
```
Should we remove this?

enejb replied:
For some reason I thought it was needed but after testing again. it doesn't seem to be the case anymore. 🙃

tyxla commented on client/components/swipeable/style.scss:
```diff
@@ -0,0 +1,29 @@
+	position: relative;
+	width: 100%;
+	height: 100%;
+	z-index: 1;
```
I'd suggest being careful about hard-coded z-indexes. We try to make them relative throughout Calypso in order to make every element work properly across different areas of Calypso. They're managed in a centralized manner too, see https://github.com/Automattic/wp-calypso/blob/trunk/client/assets/stylesheets/shared/functions/_z-index.scss

enejb replied:
Good call. I remove the line.

tyxla reviewed (approved):
Not much to add on top of @ciampo's extensive review! Just several nits and minor suggestions.

Nice work!

arthur791004 commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,296 @@
+		const targetHeight = pagesRef.current?.querySelector( '.is-current' )?.offsetHeight;
+
+		if ( targetHeight && pagesStyle?.height !== targetHeight ) {
+			height = { height: null };
```
Why do we set the height to null when the height is changed? This one makes the height animation doesn't work and the height of the container would be the max one of the children.

enejb replied:
You are right. This is currently not working anymore as expected.

arthur791004 commented:
Making the Swipeable component is great! But now if the user is swiping to the next page, they can only see the empty content. It might be better if we can show the content of the next page when the user is swiping to the next page.

arthur791004 commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,296 @@
+			const absoluteDelta = Math.abs( delta );
+			const velocity = absoluteDelta / ( dragPosition.timeStamp - dragStartData.timeStamp );
+
+			const hasMetThreshold = absoluteDelta > OFFSET_THRESHOLD || velocity > VELOCITY_THRESHOLD;
```
I'm not sure but I think it might be better to make `OFFSET_THRESHOLD` be about 40% ~ 50% of the prev/next page width.

enejb replied:
> I think it might be better to make OFFSET_THRESHOLD be about 40% ~ 50% of the prev/next page width.

Can you explain a bit more? 

The intend right now is unless you scroll far enough then will eventually trigger the slide.

arthur791004 replied:
I mean lots of carousels libarry use the percentage as the threshold. If the scroll percentage doesn't over about 40%~50%, it won't switch to the prev/next page. But now we use fixed 100px for that. it might be slightly weird if the width of the component is a bigger value or a very small value. You can imagine that you just scroll a little or accidentally, but it switches to prev/next page immediately if the value is big. Or you cannot scroll if the width of the component is smaller than your threshold.

enejb replied:
Thanks for the explanation @arthur791004. Will see how I can adjust that part.

enejb commented:
Thanks for the feedback @arthur791004  and @tyxla !

> Making the Swipeable component is great! But now if the user is swiping to the next page, they can only see the empty content. It might be better if we can show the content of the next page when the user is swiping to the next page.

I think this would be a great addition to the component but in order to keep things simpler and shippable I want to keep this feature out of the current PR. 

If we come across this requirement ( Right now I am just trying to make things work as they on the home section. 
the slider is not able to move the right indefinitely. )

enejb committed:
Remove the z-index

enejb committed:
Update the int to number on the readme

enejb committed:
Update to use optional caining

enejb committed:
Update example to be a bit more useful.

enejb committed:
Fix the height not adjusting as expected any more.

enejb committed:
More code clean up.

enejb committed:
Remove not needed code

enejb commented:
@tyxla, @arthur791004 and @ciampo 
I changed quite a bit here can you test it once again? 

Thank you so much!

enejb committed:
Reset the yarn.lock after a bad rebase

enejb committed:
Make the width treshold & based instead of a fixed value.

tyxla reviewed (commented):
I think this looks and works great! The only feedback I have is [this one](https://github.com/Automattic/wp-calypso/pull/58690#discussion_r767551141).

ciampo commented on client/components/swipeable/index.js:
```diff
@@ -0,0 +1,290 @@
+
+	const getOffset = useCallback(
+		( index ) => {
+			const offset = pageWidth * index;
```
`pageWidth` starts as `null`, and could be technically also set to `undefined` when calling `setPageWidth( resizeObserverRef.current?.getBoundingClientRect().width );`

Therefore, it would be best to check for it not to be `null` or `undefined` before using it a as a number value. The same check should be performed also before using `pageWidth` in `handleDragEnd`.

ciampo reviewed (approved):
Changes look good! I can only spot a couple of comments that haven't been addressed:

- [this minor comment](https://github.com/Automattic/wp-calypso/pull/58690#discussion_r762950062) about adding left/right arrows in the examples
- [this comment](https://github.com/Automattic/wp-calypso/pull/58690#discussion_r762160228) about refactoring the callback passed to `useUpdateLayout` 

I also left another comment, but overall all of the comments that are left are minor, and I wouldn't consider them as blockers.

enejb committed:
Use ResizeObserver instead of resize and orientationchange listeners

enejb commented:
Thanks for the feedback @tyxla  and @ciampo 
I manged to now use the ` Resize Observe` api instead of the resize and orientationchange listeners. 

I think now things work even nicer. Do you mind taking another look? 

@ciampo I addressed the comment in https://github.com/Automattic/wp-calypso/pull/58690#discussion_r768043245 By updating the swipeabel example to include wording that lets the user know to swipe left or right.

tyxla commented on client/components/swipeable/index.js:
```diff
@@ -1,42 +1,31 @@
 
 const OFFSET_THRESHOLD_PERCENTAGE = 0.45; // Percentage of width to travel before we trigger the slider to move to the desired slide.
 const VELOCITY_THRESHOLD = 0.5; // Speed of drag above, before we trigger the slider to move to the desired slide.
+const TRANSITION_DURATION = `300ms`;
```
Nit: no need for this to be in a string literal; it can just be a simple string.

enejb replied:
fixed

tyxla commented on client/components/swipeable/index.js:
```diff
@@ -73,52 +62,46 @@ export const Swipeable = ( {
 			setSwipeableArea( pagesRef.current?.getBoundingClientRect() );
 			setDragStartData( position );
-			setPageWidth( resizeObserverRef.current?.getBoundingClientRect().width );
+			// setPageWidth( resizeObserverRef.current?.getBoundingClientRect().width );
```
This can be removed.

enejb replied:
Good catch!

tyxla reviewed (approved):
Looks good and works well from my perspective. Nice work @enejb 🙌

enejb committed:
Code clean up
